File: DesignPatterns/01_GOF23/DP_examples/Lua/03_Behavioral/Memento/Originator.lua
local Memento = require("Memento")

-- 发起人类
local Originator = {}
Originator.__index = Originator

function Originator:new(s1, s2, s3)
    local o = {}
    setmetatable(o, self)
    o.state1 = s1
    o.state2 = s2
    o.state3 = s3
    return o
end

function Originator:ConsoleState()
    print("State_1: " .. self.state1)
    print("State_2: " .. self.state2)
    print("State_3: " .. self.state3)
end

function Originator:CreateMemento(version)
    local memento = Memento:new(version)
    memento:SetState("state1", self.state1)
    memento:SetState("state2", self.state2)
    memento:SetState("state3", self.state3)
    return memento
end

function Originator:SetMemento(m)
    self.state1 = m:GetState("state1")
    self.state2 = m:GetState("state2")
    self.state3 = m:GetState("state3")
end

return Originator

File: DesignPatterns/01_GOF23/DP_examples/Python/03_Behavioral/Memento/originator.py
from memento import Memento

# 发起人类
class Originator:
    def __init__(self, s1, s2, s3):
        self.state1 = s1
        self.state2 = s2
        self.state3 = s3
    
    def console_state(self):
        print(f"State_1: {self.state1}")
        print(f"State_2: {self.state2}")
        print(f"State_3: {self.state3}")
    
    def create_memento(self, version):
        memento = Memento(version)
        memento.set_state("state1", self.state1)
        memento.set_state("state2", self.state2)
        memento.set_state("state3", self.state3)
        return memento
    
    def set_memento(self, m):
        self.state1 = m.get_state("state1")
        self.state2 = m.get_state("state2")
        self.state3 = m.get_state("state3")
